File: examples/exchange/service.py
import netsvc
import sys
import signal
import string
import random

count = len(sys.argv)-1
server = None
domain = ""

args = sys.argv[1:]

def usage():
  print "Usage: service -connect server [ -domain name ]"

try:
  i = 0
  while i < count:
    if args[i] == "-connect":
      server = args[i+1]
      i = i + 2
    elif args[i] == "-domain":
      domain = args[i+1]
      i = i + 2
    else:
      usage()
      sys.exit(1)
except:
  usage()
  sys.exit(1)

if server == None:
  usage()
  sys.exit(1)

class Publisher(netsvc.Service):
  
  def __init__(self):
    netsvc.Service.__init__(self,"service","*",netsvc.SERVICE_HIDDEN)
    self.subscribeStartup(self.startup)

  def startup(self):
    self.exportMethod(self.multiply)
    self.exportMethod(self.wait)
    self.exportMethod(self.echo)
    self.announceAvailability()
    self.joinGroup("group")
    self.publishReport("creationTime",netsvc.DateTime(),-1)
    self.startTimer(self.publish,1.0)

  def publish(self,name):
    self.publishReport("currentTime",netsvc.DateTime())
    self.startTimer(self.publish,1.0)

  def multiply(self,x,y):
    return x*y

  def wait(self,period):
    self.startTimer(self.resumeResponse,period,self.conversationId())
    self.suspendResponse()

  def echo(self,value):
    return value

dispatcher = netsvc.Dispatcher()
dispatcher.monitor(signal.SIGINT)

exchange = netsvc.Exchange(netsvc.EXCHANGE_CLIENT,domain)

delay = 5
host,port = string.splitfields(server,':')
port = int(port)
exchange.connect(host,port,delay)

publisher = Publisher()
publisher.scheduleStartup()

dispatcher.run()

File: examples/exchange/service.cc
/*
// ============================================================================
//
// = FILENAME
//     service.cc
//
// = AUTHOR(S)
//     Graham Dumpleton
// 
// = COPYRIGHT
//     Copyright 2001-2004 DUMPLETON SOFTWARE CONSULTING PTY LIMITED
//
// ============================================================================
*/

#include <ONS/agent/service.hh>
#include <OTC/message/exchange.hh>
#include <OTC/dispatch/dispatch.hh>
#include <OTC/dispatch/shutdown.hh>
#include <OTC/system/program.hh>

#include <signal.h>

/* ------------------------------------------------------------------------- */

class Service : public ONS_Service
{
  public:

    			Service();

    void		startup();

    void		publish();

    double		multiply(double theValue1, double theValue2);

    void		wait(double thePeriod);

    void		echo(OTC_ROPayload thePayload);
};

Service::Service()
  : ONS_Service("service","*",OTCLIB_SERVICE_HIDDEN)
{
  subscribeStartup(&Service::startup);
}

void Service::startup()
{
  exportMethod(&Service::multiply,"multiply");
  exportMethod(&Service::wait,"wait");
  exportMethod(&Service::echo,"echo");

  announceAvailability();

  publishReport("creationTime",OTC_Time::currentTime(),-1);

  startTimer(&Service::publish,1.0);
}

void Service::publish()
{
  publishReport("currentTime",OTC_Time::currentTime(),-1);

  startTimer(&Service::publish,1.0);
}

double Service::multiply(double theValue1, double theValue2)
{
  return theValue1 * theValue2;
}

void Service::wait(double thePeriod)
{
  startTimer(&ONS_Service::resumeResponse,thePeriod,conversationId());
  suspendResponse();
}

void Service::echo(OTC_ROPayload thePayload)
{
  returnResponse(thePayload);
}

void usage()
{
  cerr << "Usage: " << OTC_Program::basename();
  cerr << " -connect server [ -domain name ]" << endl;
}

int main(int argc, char* argv[])
{
  // Save away program arguments.

  OTC_Program::initialise(argc,argv);

  // Decode the arguments.

  OTC_String theServerString;
  OTC_String theDomainName;

  while (OTC_Program::numOptions() != 0)
  {
    if (OTC_Program::option(1) == "-connect")
    {
      OTC_Program::shift(1);
      if (OTC_Program::numOptions() == 0)
	break;

      theServerString = OTC_Program::option(1);
    }
    else if (OTC_Program::option(1) == "-domain")
    {
      OTC_Program::shift(1);
      if (OTC_Program::numOptions() == 0)
	break;

      theDomainName = OTC_Program::option(1);
    }
    else
    {
      usage();
      return 1;
    }

    OTC_Program::shift(1);
  }

  if (theServerString.isEmpty())
  {
    usage();
    return 1;
  }

  // Initialise the dispatcher.

  OTC_Dispatcher::initialise();

  // Monitor signals for shutdown.

  OTCEV_Shutdown::monitor(SIGINT);

  // Connect to the exchange process.

  OTC_String theHostname;
  int thePort = 0;

  OTC_Exchange theExchange(OTCLIB_EXCHANGE_CLIENT,theDomainName);

  char const* theHostPort;
  theHostPort = theServerString.string();

  while (*theHostPort != EOS && *theHostPort != ':')
    theHostname += *theHostPort++;

  if (*theHostPort == ':')
    theHostPort++;

  if (sscanf(theHostPort,"%d",&thePort) != 1)
    thePort = 0;

  if (theHostname.isEmpty() || thePort == 0)
  {
    cerr << OTC_Program::basename() << ": invalid server" << endl;
    return 1;
  }

  theExchange.connect(theHostname,thePort,1000);

  // Create publisher.

  Service theService;
  theService.scheduleStartup();

  // Run the dispatcher.

  OTC_Dispatcher::run();

  return 0;
}

/* ------------------------------------------------------------------------- */
